main.go

```diff
@@ -7,28 +7,73 @@ import (
 	"log"
 	"net/http"
 	"os"
+	"sync"
 	"time"
 
 	"github.com/avast/retry-go"
 )
 
 const (
-	defaultRetries = 3
-	defaultTimeout = 45 * time.Second
+	defaultTimeout                       = 90 * time.Second
+	defaultMaxParallelDownloadsPerServer = 8
+	defaultMaxRetries                    = 5
+	defaultChunkSize                     = 8192
+	defaultWaitAfterError                = 0 * time.Minute
 )
 
-type downloader struct {
-	client  *http.Client
-	retries uint
+type DownloadStatus struct {
+	URL                 string
+	DownloadedFilePath  string
+	FileSizeBytes       uint64
+	DownloadedFileBytes uint64
+	Error               error
 }
 
-func (d *downloader) downloadWithContext(ctx context.Context, u string) ([]byte, error) {
+func (s *DownloadStatus) IsFinished() bool {
+	return s.Error == nil && s.DownloadedFileBytes == s.FileSizeBytes
+}
+
+type Downloader struct {
+	Client                        *http.Client
+	Timeout                       time.Duration
+	MaxParallelDownloadsPerServer uint
+	MaxRetries                    uint
+	ChunkSize                     uint64
+	WaitAfterError                time.Duration
+}
+
+func (d *Downloader) SetTimeout(t time.Duration) *Downloader {
+	d.Timeout = t
+	d.Client.Timeout = t
+	return d
+}
+func (d *Downloader) SetMaxParallelDownloadsPerServer(n uint) *Downloader {
+	d.MaxParallelDownloadsPerServer = n
+	return d
+}
+
+func (d *Downloader) SetMaxRetries(n uint) *Downloader {
+	d.MaxRetries = n
+	return d
+}
+
+func (d *Downloader) SetChunkSize(n uint64) *Downloader {
+	d.ChunkSize = n
+	return d
+}
+
+func (d *Downloader) SetWaitAfterError(t time.Duration) *Downloader {
+	d.WaitAfterError = t
+	return d
+}
+
+func (d *Downloader) downloadFileWithContext(ctx context.Context, u string) ([]byte, error) {
```

Comment:
Exportar e manter a mesma API? `DownloadWithContext(urls ...string) Result...?

Reply:
Tu tem algum caso de uso em mente para isso? Pergunto pq eu vejo isso mais como um “detalhe de implementação” que eu gostaria de ter flexibilidade para mudar (por exemplo, quando implementarmos HTTP content range).

Reply:
Eu acho que é uma boa prática, sempre que possível, exportar uma API com contexto. Lembrando que, para além de timeouts ou deadline, contextos são uma forma de troca de informações request-scoped (mais detalhes [aqui](https://vivek-kaushal.medium.com/what-really-is-golangs-context-b51874809f0f) e [aqui](https://pkg.go.dev/golang.org/x/net/context)).

Mas entendo que, caso a ideia seja uma API bem simples, esconder pode ser uma opção.

Reply:
Entendi (mesmo sem ler as referências ainda, vou ler mais tare).

Mas eu acho que essa função específica é mais um detalhe de implementação mesmo. Talvez a gente queira exportar uma `DownloadWithContext` (similar à [`Download`](https://github.com/cuducos/chunk/blob/13d640fef632d0cc0962c6db8f7fc60778fafb77/main.go#L112), e não essa `downloadWithContext` que cuida de um único arquivo).

Digo isso pois quando implementarmos o HTTP range request, vejo grande chance dessa função `downloadWithContext` virar `downloadChuckWithContext`, algo que não teria muita utilidade pro mundo externo.

Reply:
Entendi. Faz total sentido!

Reply:
Massa! Vou implementar um protótipo do `DownloadWithContext` então : )

Reply:
O que acha do 6f5a4df?

A ideia era levar o contexto passado pelo usuário para o mesmo escopo do nosso, e lá naquele  `select` capturar o cancelamento vindo do usuátrio para propagar… faz sentido?